scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/TimeUtility.java

```diff
@@ -0,0 +1,118 @@
+package gov.nasa.jpl.aerie.scheduler;
```

Comment:
A more general question do we not have a Time calculation class that does time conversion and math?

Reply:
I do not think so. I think we would benefit from having a central place for these things. Is there a module that all other already depend on or could depend on? Also, do you think there would be something wrong with these robust-to-overflow operations moving inside Duration?

Reply:
`Duration` is part of the `merlin.protocol` package, which means it's part of the data-carrying interface between the simulation driver and a model. I'd rather not upstream scheduler utilities into a package that everyone else is forced to depend on.

Reply:
Arguably, the `Duration` in `merlin.protocol` should be significantly reduced in scope to begin with. If it's spread far beyond its purpose as a data carrier between driver and model, then it's too attractive as a coupling point, and we should try to minimize that potential.

Reply:
I see. Note that I was only thinking about pulling the "bounded" operations to the `Duration` class, not centralize conversions between "string" time and `Duration`. Another class would do that.

Reply:
Yes, fair.

It's not expected that modelers will need to worry about the end of time (so to speak); if a simulation spans 600,000 years, an `ArithmeticException` is likely just a symptom of a deeper problem. `Duration` does not even provide an absolute frame of reference; most uses between the model and driver are relative (e.g. `delay(2, SECONDS)`.

I suspect that, if the scheduler has legitimate need for saturating operations on durations, it's confronting time with a meaningfully different attitude than is meant to be served by the `merlin.protocol` `Duration`.

Reply:
(Correct me if I'm wrong, Adrien) - I think we're most likely to run into overflow issues where we denote "inifinity" by "Duration.MAX_VALUE". For example, in `Window`:
https://github.com/NASA-AMMOS/aerie/blob/45d42efc91dbe676fd9aad9e620315a637c1cfe9/constraints/src/main/java/gov/nasa/jpl/aerie/constraints/time/Window.java#L125

Using all of the operations in `Window`, this poses no issue. However, the scheduler adds additional operations to shift windows forward and backward in time - which would result in an overflow.

I bet we could model our way around this without running into ArithmeticExceptions, or needing to change `Duration` directly.

(If there's another use case I'm missing here, let me know @adrienmaillard )

Reply:
That's a good point, and I think I committed the original sin there. I agree that if you're translating windows around, this representation of "forever" no longer cuts it.

Reply:
You may be right @mattdailis. The overflow especially happens with loose temporal constraints such as endsafter() or endsbefore(), startsafter(), startsbefore(). They were modelled as a minimal operation on a time point where you add/subtract infinity and get the resulting interval without really modelling the constraint explicitly. We could revise that and do differently but I'd say this is out of scope (and there is already a lot of out-of-scope stuff in this PR).

Reply:
Yes, FWIW I'm not pushing back on anything in this PR (not that I've looked!). Just against upstreaming stuff into Duration.

Reply:
Totally agree on the scope :+1: I've captured this potential design improvement in a JIRA ticket: [Model "forever" better in the scheduler](https://jira.jpl.nasa.gov/browse/AERIE-1780)

Copying the content below for the benefit of those who don't have access to JIRA (which is most people in this thread... 🙄 ):

> As of writing, the scheduler models "forever" as a window from Duration.MIN_DURATION to Duration.MAX_DURATION. This worked well for operations like intersections (defined on constraints) but it causes an arithmetic overflow for the additional operations added by the scheduler (which might look to create windows before and after another window).
>
> The goal of this ticket is to design and implement a solution that does not use MAX_VALUE and MIN_VALUE to denote infinite durations. This solution should not need to touch any classes in merlin.protocol, since this is purely a scheduling concern.